### PR TITLE
New model node: YOLOX

### PR DESCRIPTION
**Pull request by liyier90**
Closes https://github.com/aimakerspace/PeekingDuck-Private/issues/12
- [x] [PDR](https://www.notion.so/34965a02a50b4353865ce3e61449d10d?v=e1c750855b144f26aa3420a7be49e2f9)
- [x] [CDR](https://www.notion.so/d5af8803446f471488230287f849b2cc?v=719a3313d81147629af2c47d96cd345e)
- [ ] Speed/Accuracy benchmarks

**Changes:**
- Implemented `model.yolox`
- Refactored away some unused layers
  - Original project provided the option to use depth-wise conv (`DWConv`) but provided pre-trained weights do not use it
  - Since we are not training our own general object detector, code related to `DWConv` has been removed
  - If `DWConv` is found to be crucial for the performance of future custom trained models, it can be added again.
- Similar for activation functions, only `SiLU` is used by the pre-trained weights
- Added utility functions to convert bbox to PKD's format
- Added utility script to strip optimizer data from the weights (not covered by unit tests)

**Config options:**
Only listing the ones that are not common/found in other existing models
- fp16: Flag to determine if half-precision floating-point should be used
            for inference.
- fuse : Flag to determine if the convolution and batch normalization layers
            should be fused for inference.

**Tests:**
- Close to 100% coverage in [Post Merge Checks workflow](https://github.com/liyier90/PeekingDuck/runs/4350351171?check_suite_focus=true)

| Name | Stmts | Miss | Cover |
|------|-------|------|-------|
| peekingduck/pipeline/nodes/model/yolox.py | 12 | 0 | 100% |
| peekingduck/pipeline/nodes/model/yoloxv1/\_\_init__.py | 1 | 0 | 100% |
| peekingduck/pipeline/nodes/model/yoloxv1/yolox_files/\_\_init__.py | 1 | 0 | 100% |
| peekingduck/pipeline/nodes/model/yoloxv1/yolox_files/darknet.py | 33 | 0 | 100% |
| peekingduck/pipeline/nodes/model/yoloxv1/yolox_files/detector.py | 79 | 2 | 97% |
| peekingduck/pipeline/nodes/model/yoloxv1/yolox_files/model.py | 105 | 0 | 100% |
| peekingduck/pipeline/nodes/model/yoloxv1/yolox_files/network_blocks.py | 65 | 0 | 100% |
| peekingduck/pipeline/nodes/model/yoloxv1/yolox_files/utils.py | 37 | 0 | 100% |
| peekingduck/pipeline/nodes/model/yoloxv1/yolox_model.py | 36 | 0 | 100% |
- 97% coverage of `detector.py` is due to lines that only gets called when GPU is enabled, 100% when running on a GPU enabled machine locally. Detailed explanation can be found in the the CDR.

*Note:*
- The file size of the uploaded weights can potentially be reduced if half-precision weights can give the same accuracy
  - Pending mAP evaluation
- I merged the latest `dev` by mistake and introduced a bunch of extra commits. If it's preferred, I can create a new branch and put the changed files there. Caveat, it will be just a single commit.

### CODE REVIEW

**Comment by ongtw** [comment]
Nice! Looking forward to replacing Yolo-v4-tiny with YoloX.

File: tests/pipeline/nodes/model/yoloxv1/test_yolox.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+        npt.assert_equal(output["bboxes"], expected_output["bboxes"])
+        npt.assert_equal(output["bbox_labels"], expected_output["bbox_labels"])
+        npt.assert_equal(output["bbox_scores"], expected_output["bbox_scores"])
+
+    def test_return_at_least_one_person_and_one_bbox(self, test_human_images, yolox):
+        test_image = cv2.imread(test_human_images)
+        output = yolox.run({"img": test_image})
+
+        assert "bboxes" in output
+        assert output["bboxes"].size > 0
```
Comment (ongtw) [inline tests/pipeline/nodes/model/yoloxv1/test_yolox.py:125]: Current testing only checks bboxes structure (size). It would be more complete and robust to check the actual bboxes contents. E.g. if the model erroneously returns a bbox with (0,0,0,0), it will pass the size check but the contents are invalid.
The purpose of doing a (predetermined) content test is to identify the scenario whereby a developer makes a future change to the model and accidentally corrupts the internal contents.

**Review by ongtw — COMMENTED** [review]
Some comments for your consideration, please.

File: peekingduck/pipeline/nodes/model/yoloxv1/yolox_files/darknet.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+    # pylint: disable=arguments-differ
+    def __init__(
+        self,
+        dep_mul: float,
+        wid_mul: float,
+        out_features: Tuple[str, str, str],
+    ) -> None:
+        super().__init__()
+        assert out_features, "please provide output features of Darknet"
```
[the remark this answers is not in the document]
Comment (liyier90) [reply peekingduck/pipeline/nodes/model/yoloxv1/yolox_files/darknet.py]: Removed unnecessary `assert`

File: peekingduck/pipeline/nodes/model/yoloxv1/yolox_files/darknet.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+class CSPDarknet(nn.Module):
+    """Modified CSPNet with SiLU activation."""
+
+    # pylint: disable=arguments-differ
+    def __init__(
+        self,
+        dep_mul: float,
+        wid_mul: float,
+        out_features: Tuple[str, str, str],
+    ) -> None:
```
[the remark this answers is not in the document]
Comment (liyier90) [reply peekingduck/pipeline/nodes/model/yoloxv1/yolox_files/darknet.py:70]: Added documentation to explain how the arguments are used

File: peekingduck/pipeline/nodes/model/yoloxv1/yolox_files/darknet.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        self.dark4 = nn.Sequential(*self.make_group_layer(channels * 4, depth * 3))
+        self.dark5 = nn.Sequential(
+            BaseConv(channels * 8, channels * 16, 3, 2),
+            SPPBottleneck(channels * 16, channels * 16),
+            CSPLayer(channels * 16, channels * 16, depth, False),
+        )
+
+    @staticmethod
+    def make_group_layer(in_channels: int, depth: int) -> Tuple[BaseConv, CSPLayer]:
+        """Starts with BaseConv layer, followed by a CSPLayer"""
```
[the remark this answers is not in the document]
Comment (liyier90) [reply peekingduck/pipeline/nodes/model/yoloxv1/yolox_files/darknet.py]: Converted to Google-format docstring, similar for `forward()`

File: peekingduck/pipeline/nodes/model/yoloxv1/yolox_files/detector.py [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+        Args:
+            prediction (torch.Tensor): Raw detections from the YOLOX model.
+            scale (float): Scale factor used during preprocessing.
+            image_shape (Tuple[int, int]): Image size of the original input
+                image.
+            class_names (List[str]): List of human-friendly object catetory
+                names.
+
+        Returns:
+            (Tuple[List[np.ndarray], List[str], List[float]]): Returned tuple
```
[the remark this answers is not in the document]
Comment (liyier90) [reply peekingduck/pipeline/nodes/model/yoloxv1/yolox_files/detector.py:180]: Parenthesis is used to indicate that content is a type similar to how types are indicated/represented in `Args`. I can remove it if you feel that it's unnecessary.

File: peekingduck/pipeline/nodes/model/yoloxv1/yolox_files/detector.py [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+            f"Input resolution: {self.input_size}\n\t"
+            f"IDs being detected: {self.config['detect_ids']}\n\t"
+            f"IOU threshold: {self.config['iou_threshold']}\n\t"
+            f"Score threshold: {self.config['score_threshold']}\n\t"
+            f"Half-precision floating-point: {self.config['half']}\n\t"
+            f"Fuse convolution and batch normalization layers: {self.config['fuse']}"
+        )
+        return self._load_yolox_weights(model_path, model_size)
+
+    def _get_model(self, model_size: Dict[str, float]) -> YOLOX:
```
[the remark this answers is not in the document]
Comment (liyier90) [reply peekingduck/pipeline/nodes/model/yoloxv1/yolox_files/detector.py:125]: Added docstring

File: peekingduck/pipeline/nodes/model/yoloxv1/yolox_files/detector.py [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+        class_names: List[str],
+    ) -> Tuple[List[np.ndarray], List[str], List[float]]:
+        """Postprocesses detection result to be compatible with other nodes.
+
+        Args:
+            prediction (torch.Tensor): Raw detections from the YOLOX model.
+            scale (float): Scale factor used during preprocessing.
+            image_shape (Tuple[int, int]): Image size of the original input
+                image.
+            class_names (List[str]): List of human-friendly object catetory
```
[the remark this answers is not in the document]
Comment (liyier90) [reply peekingduck/pipeline/nodes/model/yoloxv1/yolox_files/detector.py]: Corrected typo

File: peekingduck/pipeline/nodes/model/yoloxv1/yolox_files/detector.py [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+
+        return bboxes, classes, scores
+
+    def _preprocess(self, image: np.ndarray) -> Tuple[np.ndarray, float]:
+        """Preprocesses the input image for inference.
+
+        The input image is resized to fit the `input_size` specified in the
+        configurations. If the image cannot be resized exactly, a gray canvas,
+        rgb(114, 114, 114), is created and the resized image is pasted on the
+        top left corner of the canvas.
```
[the remark this answers is not in the document]
Comment (liyier90) [reply peekingduck/pipeline/nodes/model/yoloxv1/yolox_files/detector.py]: Added explanation that (114, 114, 114) is chosen because it is the value used by the original repo during training

File: peekingduck/pipeline/nodes/model/yoloxv1/yolox_files/model.py [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+            IN_CHANNELS[0], IN_CHANNELS[1], n_bottleneck, width
+        )
+
+        # bottom-up conv
+        self.bu_conv1 = BaseConv(
+            int(IN_CHANNELS[1] * width), int(IN_CHANNELS[1] * width), 3, 2
+        )
+        self.C3_n4 = self.make_csp_layer(
+            IN_CHANNELS[1], IN_CHANNELS[2], n_bottleneck, width
+        )
```
[the remark this answers is not in the document]
Comment (liyier90) [reply peekingduck/pipeline/nodes/model/yoloxv1/yolox_files/model.py]: Refactored with `N256, N512, N1024 = IN_CHANNELS` for clarity.

File: peekingduck/pipeline/nodes/model/yoloxv1/yolox_files/model.py [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+
+    @staticmethod
+    def make_csp_layer(
+        in_channel: int, out_channel: int, depth: int, width: float
+    ) -> CSPLayer:
+        """Returns a CSPLayer."""
```
[the remark this answers is not in the document]
Comment (liyier90) [reply peekingduck/pipeline/nodes/model/yoloxv1/yolox_files/model.py]: Converted docstrings to Google format

File: peekingduck/pipeline/nodes/model/yoloxv1/yolox_files/utils.py [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+def xywh2xyxy(inputs: torch.Tensor) -> torch.Tensor:
+    """Converts from [x, y, w, h] to [x1, y1, x2, y2] format.
+
+    (x1, y1) is the top left corner and (x2, y2) is the bottom right corner.
+    """
+    outputs = torch.empty_like(inputs)
+    outputs[:, 0] = inputs[:, 0] - inputs[:, 2] / 2
+    outputs[:, 1] = inputs[:, 1] - inputs[:, 3] / 2
+    outputs[:, 2] = inputs[:, 0] + inputs[:, 2] / 2
+    outputs[:, 3] = inputs[:, 1] + inputs[:, 3] / 2
```
[the remark this answers is not in the document]
Comment (liyier90) [reply peekingduck/pipeline/nodes/model/yoloxv1/yolox_files/utils.py:147]: Added explanation

File: requirements.txt [hunk 12]
```diff
@@ -1,9 +1,11 @@
...
 pyyaml >= 5.3.1
 requests == 2.24.0
 tensorflow == 2.2.0
+torch == 1.10.0
+torchvision == 0.11.1
 tqdm == 4.45.0
-click == 7.1.2
-scipy
\ No newline at end of file
+scipy
```
[the remark this answers is not in the document]
Comment (liyier90) [reply requirements.txt:11]: Updated M1 Mac installation instructions https://github.com/aimakerspace/PeekingDuck/pull/544/commits/458d4b95c60963951e9dfd338539a1d9d564d563

File: peekingduck/pipeline/nodes/model/yoloxv1/yolox_model.py [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+            config["root"], config["weights"], config["weights_parent_dir"]
+        )
+        if not checker.has_weights(weights_dir, model_dir):
+            self.logger.warning("No weights detected. Proceeding to download...")
+            downloader.download_weights(weights_dir, config["weights"]["blob_file"])
+            self.logger.info(f"Weights downloaded to {weights_dir}.")
+
+        with open(model_dir / config["weights"]["classes_file"]) as infile:
+            self.class_names = [line.strip() for line in infile.readlines()]
+        self.detect_ids = config["detect_ids"]
```
[the remark this answers is not in the document]
Comment (liyier90) [reply peekingduck/pipeline/nodes/model/yoloxv1/yolox_model.py]: Updated to make changes to the `detect_ids` property in `YOLOXModel` update the `detect_ids` member variable in YOLOX `Detector` as well. Sorry, missed this initially. 

`detect_ids` property `YOLOXModel` is meant return the list so user can debug or do other visualizing/processing steps more easily. `detect_ids` member variable in YOLOX `Detector` is meant to be a `torch.Tensor` so we can apply the `torch.isin()` filter more easily. 

Some attempts and failures:
- Tried to implement the `detect_ids` at YOLOX `Node`
  - AttributeError due to how `ConfigLoader` loads config for the `Node`.
- Tried initializing `detector_ids` before `detector` in `YOLOXModel.__init__` to avoid repeating calls to `update_detect_ids()` in YOLOX `Detector`
  - AttributeError since detector is not initialized yet.

File: tests/pipeline/nodes/model/yoloxv1/test_yolox.py (hunk 1 above)
Comment (liyier90) [reply tests/pipeline/nodes/model/yoloxv1/test_yolox.py:128]: Added tests to test the values of `bboxes`, bbox_labels`, `bbox_scores` from human detection results.

File: peekingduck/pipeline/nodes/model/yoloxv1/yolox_files/detector.py [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+    """
+
+    def __init__(self, config: Dict[str, Any], model_dir: Path) -> None:
+        self.logger = logging.getLogger(__name__)
+
+        self.config = config
+        self.model_dir = model_dir
+        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+        # Half-precision only supported on CUDA
+        self.config["half"] = self.config["half"] and self.device.type != "cpu"
```
[the remark this answers is not in the document]
Comment (ongtw) [reply peekingduck/pipeline/nodes/model/yoloxv1/yolox_files/detector.py]: Code updated by Yier

File: peekingduck/pipeline/nodes/model/yoloxv1/yolox_files/network_blocks.py [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+    ) -> None:
+        super().__init__()
+        hidden_channels = int(out_channels * expansion)
+        self.conv1 = BaseConv(in_channels, hidden_channels, 1, 1)
+        self.conv2 = BaseConv(hidden_channels, out_channels, 3, 1)
+        self.use_add = shortcut and in_channels == out_channels
+
+    def forward(self, inputs: torch.Tensor) -> torch.Tensor:
+        """Defines the computation performed at every call."""
+        ouputs = self.conv2(self.conv1(inputs))
```
[the remark this answers is not in the document]
Comment (liyier90) [reply peekingduck/pipeline/nodes/model/yoloxv1/yolox_files/network_blocks.py]: Fixed typo

File: peekingduck/pipeline/nodes/model/yoloxv1/yolox_files/detector.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ongtw) [reply peekingduck/pipeline/nodes/model/yoloxv1/yolox_files/detector.py:200]: Ok got your point, let's retain the current convention.
Apologies, I was confused visually by the lack of return variable name.

File: peekingduck/pipeline/nodes/model/yoloxv1/yolox_files/network_blocks.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (ongtw) [reply peekingduck/pipeline/nodes/model/yoloxv1/yolox_files/network_blocks.py]: Correction pending.

File: requirements.txt (hunk 12 above)
[the remark this answers is not in the document]
Comment (ongtw) [reply requirements.txt:11]: Do we need to specify a version for scipy?

[the remark this answers is not in the document]
Comment (liyier90) [reply requirements.txt]: Specified version for `scipy`. Used `>=` instead of `==` (pinning) following advice from https://stackoverflow.com/questions/28509481

[the remark this answers is not in the document]
Comment (ongtw) [reply requirements.txt]: The stackoverflow link is very informative.
